File: backend/routes/billing.ts
import express from 'express';

const router = express.Router();
const billingController = require('../controllers/billingController');
const isAuthenticated = require('../middlewares/isAuthenticated');
const { z } = require('zod');

const createCheckoutSchema = z.object({
  plan: z.enum(['pro', 'team']),
  interval: z.enum(['monthly', 'yearly']),
});

const activateLicenseSchema = z.object({
  license_key: z.string().min(1),
});

const validateBody = (schema: any) => (req: express.Request, res: express.Response, next: express.NextFunction) => {
  const result = schema.safeParse(req.body);
  if (!result.success) {
    return res.status(400).json({ error: 'Invalid payload', details: result.error.errors });
  }
  req.body = result.data;
  return next();
};

// Stripe webhook — raw body, no auth, no CSRF
router.post('/api/billing/webhook', billingController.handleWebhook);

// Authenticated billing routes
router.post('/api/billing/create-checkout-session', isAuthenticated, validateBody(createCheckoutSchema), billingController.createCheckoutSession);
router.post('/api/billing/create-portal-session', isAuthenticated, billingController.createPortalSession);
router.get('/api/billing/subscription', isAuthenticated, billingController.getSubscription);
router.post('/api/billing/activate-license', isAuthenticated, validateBody(activateLicenseSchema), billingController.activateLicense);

export default router;

File: backend-rust/src/routes/billing.rs
use crate::{handlers, middleware::auth::require_auth, state::AppState};
use axum::{
  middleware,
  routing::{get, post},
  Router,
};

pub fn router(state: AppState) -> Router<AppState> {
  // Webhook route — raw body, no auth, no CSRF
  let public = Router::new()
    .route("/api/billing/webhook", post(handlers::billing::handle_webhook));

  let protected = Router::new()
    .route(
      "/api/billing/create-checkout-session",
      post(handlers::billing::create_checkout_session),
    )
    .route(
      "/api/billing/create-portal-session",
      post(handlers::billing::create_portal_session),
    )
    .route(
      "/api/billing/subscription",
      get(handlers::billing::get_subscription),
    )
    .route(
      "/api/billing/activate-license",
      post(handlers::billing::activate_license),
    )
    .route_layer(middleware::from_fn_with_state(state, require_auth));

  public.merge(protected)
}
